### PR TITLE
torch: forbid using skip_synchronize in async training

### PR DESCRIPTION
**Pull request by pleasantrabbit**
After optimizer.skip_synchronization() is called, optimizer.step()
returns after triggering push-pull. Otherwise the local gradients from
the current pass will be lost.

Fix license classifier.
Add OS classifier.

### CODE REVIEW

**Comment by ymjiang** [comment]
@pleasantrabbit Thank you for reporting this bug. Indeed, https://github.com/bytedance/byteps/pull/236 has a bug in async mode. 

I feel like `skip_synchronize` should not be used in async training because users will manually call `optimizer.synchronize()`, which is incompatible with current async design. 

Therefore, I added a new commit that checks `skip_synchronize` is only used in synchronous mode. @bobzhuyb What do you think about this?
